File: backend/app/schemas/driver.py
"""Driver (Delivery Partner) Schemas."""

from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field
from app.schemas.common import VerificationStatus


class DriverAvailabilityStatus(str, Enum):
    AVAILABLE = "AVAILABLE"
    ON_JOB = "ON_JOB"
    OFFLINE = "OFFLINE"


class DriverProfileResponse(BaseModel):
    """Full driver / delivery partner profile response model."""

    id: str
    user_id: str
    driving_license_number: Optional[str] = None
    vehicle_type: Optional[str] = None
    vehicle_number: Optional[str] = None
    vehicle_capacity_kg: Optional[float] = None
    has_refrigeration: bool = False
    is_online: bool = False
    availability_status: DriverAvailabilityStatus = DriverAvailabilityStatus.AVAILABLE
    current_latitude: Optional[float] = None
    current_longitude: Optional[float] = None
    verification_status: VerificationStatus = VerificationStatus.PENDING
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class DriverProfileUpdate(BaseModel):
    """Client update payload for driver profile.
    Notice: verification_status is excluded to prevent client tampering.
    """

    vehicle_type: Optional[str] = Field(None, max_length=50)
    vehicle_number: Optional[str] = Field(None, max_length=50)
    vehicle_capacity_kg: Optional[float] = Field(None, ge=0)
    has_refrigeration: Optional[bool] = None
    is_online: Optional[bool] = None
    availability_status: Optional[DriverAvailabilityStatus] = None
    current_latitude: Optional[float] = Field(None, ge=-90.0, le=90.0)
    current_longitude: Optional[float] = Field(None, ge=-180.0, le=180.0)


class DriverLocationUpdate(BaseModel):
    """Payload for driver GPS telemetry update."""

    latitude: float = Field(..., ge=-90.0, le=90.0, description="Latitude coordinate")
    longitude: float = Field(..., ge=-180.0, le=180.0, description="Longitude coordinate")
    accuracy_meters: Optional[float] = Field(None, ge=0.0, le=1000.0, description="GPS accuracy in meters")


class DriverLocationResponse(BaseModel):
    id: str
    driver_id: str
    latitude: float
    longitude: float
    accuracy_meters: Optional[float] = None
    recorded_at: str


class EligibleDriverResponse(BaseModel):
    """Suggested eligible driver for dispatch matching."""

    id: str
    driver_user_id: str
    name: str
    phone: Optional[str] = None
    vehicle_type: str
    vehicle_number: Optional[str] = None
    vehicle_capacity_kg: float
    has_refrigeration: bool
    distance_to_pickup_km: float
    pickup_eta_minutes: int
    availability_status: DriverAvailabilityStatus
    verification_status: VerificationStatus

File: anna-setu-frontend-development/lib/api/driver.ts
/**
 * Driver (Delivery Partner) API Client
 */

import { apiClient } from './client'
import { VerificationStatus } from '@/lib/types/core'

export interface DriverProfile {
  id: string
  user_id: string
  driving_license_number?: string
  vehicle_type?: string
  vehicle_number?: string
  vehicle_capacity_kg?: number
  has_refrigeration?: boolean
  is_online?: boolean
  current_latitude?: number
  current_longitude?: number
  verification_status: VerificationStatus
  created_at?: string
  updated_at?: string
}

export interface DriverProfileUpdatePayload {
  vehicle_type?: string
  vehicle_number?: string
  vehicle_capacity_kg?: number
  has_refrigeration?: boolean
  is_online?: boolean
  current_latitude?: number
  current_longitude?: number
}

export interface DriverLocationUpdatePayload {
  latitude: number
  longitude: number
  accuracy_meters?: number
}

export interface DriverLocationRecord {
  id: string
  driver_id: string
  latitude: number
  longitude: number
  accuracy_meters?: number
  recorded_at: string
}

export const driverApi = {
  /**
   * Fetch driver profile
   */
  async getProfile(): Promise<DriverProfile> {
    return apiClient<DriverProfile>('/api/v1/driver/profile')
  },

  /**
   * Update driver operational fields
   */
  async updateProfile(payload: DriverProfileUpdatePayload): Promise<DriverProfile> {
    return apiClient<DriverProfile>('/api/v1/driver/profile', {
      method: 'PATCH',
      body: JSON.stringify(payload),
    })
  },

  /**
   * Submit driver live GPS telemetry
   */
  async updateLocation(payload: DriverLocationUpdatePayload): Promise<DriverLocationRecord> {
    return apiClient<DriverLocationRecord>('/api/v1/driver/location', {
      method: 'POST',
      body: JSON.stringify(payload),
    })
  },

  /**
   * Fetch active and completed deliveries assigned to the authenticated driver
   */
  async getMyDeliveries(): Promise<any[]> {
    return apiClient<any[]>('/api/v1/driver/deliveries')
  },
}

